### PR TITLE
Give COLR an explicit bbox to call it's own

### LINKED ISSUE #251: Consider allowing indirection to retrieve bounding box from different glyf glyph
In a conversation with @bungeman Ben points at a certain restriction when using a font as an outline font and COLRv1 font at the same time. I.e. displaying a COLRv1 font in an environment that does not understand COLRv1.

If the bounding box is always retrieved from the base glyph's entry in the `glyf` table, this fallback contour glyph would always need to be at least as wide as its COLRv1 rendering result (potentially widened by invisible points if needed).

Font makers would have more flexibility with the fallback `glyf` glyphs if we had a way to point to a dedicated bounding box glyph separately, other than always referring to the base glyph's `glyf` entry.

### PR DESCRIPTION
**Pull request by rsheeter**
Strawman for a fix for #251. Costs 16 bytes per BaseGlyphPaintRecord ... but for a font that doesn't _have_ an alternate presentation you get some of them back by not writing out bogus glyph records just to express bboxes (as our exploratory compiler does today).

Edit: Fixes #251

### CODE REVIEW

**Comment by rsheeter** [comment]
Another interesting benefit: bbox sharing Just Works, for things with repetitive bboxes (emoji, icons) that's a win.

File: OFF_AMD2_WD.md [hunk 1]
```diff
@@ -1444,6 +1442,16 @@ in COLR version 0, providing records that map a base glyph to a color glyph
...
 different, however—see 5.7.11.1.
 
+*VarBBox table:*
+
+| Type | Name | Description |
+|-|-|-|
+| UFWORD | x | x-coord of bounding box. For variation, use varIndexBase + 0. |
+| UFWORD | y | x-coord of bounding box. For variation, use varIndexBase + 1. |
+| UFWORD | w | width of bounding box. For variation, use varIndexBase + 2. |
+| UFWORD | h | height of bounding box. For variation, use varIndexBase + 3. |
```
Comment (justvanrossum) [inline OFF_AMD2_WD.md:1452]: Any particular reason to use width/height instead of max_x/max_y? Intuitively I think absolute coordinates are easier to deal with in the context of variations.

Comment (PeterConstable) [reply OFF_AMD2_WD.md:1452]: I agree: I'd use xMin, yMin, xMax, yMax to match what is done in the 'glyf' and 'head' tables.

File: OFF_AMD2_WD.md [hunk 2]
```diff
@@ -1444,6 +1442,16 @@ in COLR version 0, providing records that map a base glyph to a color glyph
 definition. The color glyph definitions that each refer to are significantly
 different, however—see 5.7.11.1.
 
+*VarBBox table:*
+
+| Type | Name | Description |
+|-|-|-|
+| UFWORD | x | x-coord of bounding box. For variation, use varIndexBase + 0. |
```
Comment (PeterConstable) [inline OFF_AMD2_WD.md:1449]: These need to be FWORD, not UFWORD.

File: OFF_AMD2_WD.md [hunk 3]
```diff
@@ -1457,6 +1465,7 @@ different, however—see 5.7.11.1.
 |-|-|-|
 | uint16 | glyphID | Glyph ID of the base glyph. |
 | Offset32 | paintOffset | Offset to a Paint table. |
+| Offset32 | bboxOffset | Offset to a VarBBox table. |
```
Comment (PeterConstable) [inline OFF_AMD2_WD.md:1468]: I'd state explicitly that bboxOffset can be NULL, in which case the bbox data is taken from the base glyph data. I don't see any downside of allowing that for cases in which that would be adequate, and it can save 12 bytes per colour glyph.

File: colr-gradients-spec.md [hunk 4]
```diff
@@ -525,10 +525,19 @@ struct PaintComposite
   Offset24<Paint>        backdrop;
 };
 
+struct VarBBox {
+  UFWORD                 x; // VarIdx varIndexBase + 0.
```
Comment (PeterConstable) [inline colr-gradients-spec.md:529]: Type should be FWORD. Change members to xMin, yMin, xMax, yMax.

**Review by PeterConstable — CHANGES_REQUESTED** [review]

File: OFF_AMD2_WD.md (hunk 2 above)
Comment (rsheeter) [reply OFF_AMD2_WD.md:1449]: Good catch, fixing

File: OFF_AMD2_WD.md (hunk 1 above)
Comment (rsheeter) [reply OFF_AMD2_WD.md:1452]: Apparently somewhere along the line x,y,w,h stuck in my head as the definition of box :) Agreed it's best to match existing practice.

File: OFF_AMD2_WD.md (hunk 3 above)
Comment (rsheeter) [reply OFF_AMD2_WD.md:1468]: I waffle a bit on that one. Thinking "aloud", two reasons to potentially not do that:

1. Simplicity. Bbox is trivially accessible for both glyf and cff with COLR.
1. Discussion on #251 where a pretty expert group of users expressed concern with use of base glyph

File: OFF_AMD2_WD.md (hunk 1 above)
Comment (rsheeter) [reply OFF_AMD2_WD.md]: Done

File: OFF_AMD2_WD.md (hunk 2 above)
Comment (rsheeter) [reply OFF_AMD2_WD.md]: Done

File: colr-gradients-spec.md (hunk 4 above)
Comment (rsheeter) [reply colr-gradients-spec.md]: Done

File: OFF_AMD2_WD.md (hunk 3 above)
Comment (rsheeter) [reply OFF_AMD2_WD.md:1468]: Would it only be true for `glyf` that bbox as described in https://docs.microsoft.com/en-us/typography/opentype/spec/glyf is sufficient or is that true of CFF as well?

Comment (PeterConstable) [reply OFF_AMD2_WD.md:1468]: >Simplicity. Bbox is trivially accessible for both glyf and cff with COLR.

Not clear to me what you're saying and why it argues against optionality.

> Discussion on [issue 251...] where a pretty expert group of users expressed concern with use of base glyph

@raphlinus brought up something that I think is a non-issue. I don't There are working implementations of COLR v1 now that have successfully used bbox info from 'glyf'. Those work without issue.

Comment (PeterConstable) [reply OFF_AMD2_WD.md:1468]: >Would it only be true for glyf that bbox as described in [OT 'glyf' spec] is sufficient or is that true of CFF as well?

The concept of the bounding box described in the 'glyf' chapter — derived from point coordinate data for on- or off-curve points — would apply equally to CFF(2) as well as to 'glyf'. However, whereas in the 'glyf' table the glyph entries include the bbox values (xMin, etc.), I don't think there's an equivalent in CFF(2) entries. So, a bbox would have to be computed by processing the charstring data.

We could say that a bbox table _should_ (or _must_ / _shall_) be used for fonts that use CFF(2) for base glyphs.

File: colr-gradients-spec.md [hunk 5]
```diff
@@ -525,10 +525,19 @@ struct PaintComposite
   Offset24<Paint>        backdrop;
 };
 
+struct VarBBox {
```
Comment (PeterConstable) [inline colr-gradients-spec.md:528]: Since the same struct will be used for both variable and non-variable cases, is "Var" needed in the name? (We add "Var" in other structs when we're distinguishing from a corresponding non-variable struct.)

Comment (rsheeter) [reply colr-gradients-spec.md:528]: I was going for consistency in that Var* would be a variable thing. Don't really have strong feelings about it.

File: OFF_AMD2_WD.md (hunk 3 above)
Comment (rsheeter) [reply OFF_AMD2_WD.md:1458]: I prefer not to have a case where you process the CFF or glyf glyph to get the box. To achieve that I believe I need  it for CFF, and I need it for a variable glyf, and I need it for a glyf with an alternate representation, so we'd be making it optional solely when  there is a glyf table and the bbox desired exists in glyf. And it should be fairly efficient given it's sharable. That leaves me thinking having it mandatory makes sense. WDYT?

File: colr-gradients-spec.md (hunk 5 above)
Comment (rsheeter) [reply colr-gradients-spec.md:525]: Per IM discussion, leaving VarBBox

**Review by PeterConstable — APPROVED** [review]

**Comment by PeterConstable** [comment]
@googlebot I consent

File: OFF_AMD2_WD.md (hunk 3 above)
Comment (rsheeter) [reply OFF_AMD2_WD.md:1458]: Resolving per IM and PR approval w/bbox mandatory

**Comment by anthrotype** [comment]
> The same structure is used in both variable and non-variable fonts.

why not having distinct BBox and VarBBox structs and save the unused 4 bytes?

**Comment by PeterConstable** [comment]
How would you determine which structure to read? Have a VarBaseGlyphPaintRecord (and whatever else that might entail)? Have the format for the (Var)BBox table conditioned by whether COLR.itemVariationStoreOffset == NULL?

**Comment by anthrotype** [comment]
ah good point, I had not thought about that. I'm not sure actually.
Checking if VarStore != NULL to determine whether to parse static BBox vs VarBBox would mean that if any color glyph has some COLR variations (and thus has a non-null VarStore), then all the color glyphs (even those that actually do not vary) will use VarBBox.
Perhaps adding a uint8 format to BaseGlyphPaintRecord that says whether it can vary or not (which selects the appropriate BBox or VarBBox), but then we'd spend 1 byte to save 4 so the net effect would be 3 bytes saved overall for static-only fonts, and +1 byte for each color glyphs that can vary in a color VF.
But I'm not sure if it's actually worth it.

**Comment by rsheeter** [comment]
After sanity checking we actually get quite a lot of unique bboxes (text fonts just for comparison, they aren't COLR):

```
twemoji-glyf_colr_1.ttf 10622 unique bboxes
noto-glyf_colr_1.ttf 39370 unique bboxes
Roboto[wdth,wght].ttf 1058 unique bboxes
Arimo[wght].ttf 2503 unique bboxes
```

Maybe worthwhile indeed.

**Comment by behdad** [comment]
I think merging this was premature.

What happens when the paint tries to paint outside the bbox? I don't see that specified. And if not specified, we are bound to see fonts that have bogus bounding box info.

**Comment by rsheeter** [comment]
I don't think this changes that concern? - we already said you should be
able to use a provided bbox, this merely changes where that lives

**Comment by behdad** [comment]
> I don't think this changes that concern? - we already said you should be able to use a provided bbox, this merely changes where that lives

What bothers me most is that this is not even optional. The byte cost cannot be avoided by an implementation that wouldn't need it.

What should have been done instead was to address Ben's concerns by specing out how to calculate bounding box on the fly. This is a very well-known problem. You simply have to run the paint graph / draw twice: first to calculate the bounds so you can allocate, second to do the drawing.

What has happened here is typical of when implementors and spec writers are too close. Things go into the spec without being fully justified. Variations, indeed, as have been pointed out multiple times, make it *impossible* to even specify bounding boxes.

Even before PaintRotate, the *tight* bounding box of an OpenType 1.8 Variable Font was *not* expressible as a variable value using the same variable model.  It *was* possible to specify a non-tight box easily-enough though.

With PaintRotate, even that is not doable. So, to address a concern about rasterization time, one that is well-understood and easily solvable, you have now encoded requirements for the compiler, that we have no idea how to implement.

Indeed, it was fine closing https://github.com/googlefonts/colr-gradients-spec/issues/289 because an implementation did NOT have to use the encoded bbox. But now we are explicitly forcing compilers to compute that, something we don't have a known way to do, short of walking every glyph outline point and calculating it's max bounds under the PaintGraph.

In short, you've replaced a tractable problem with an intractable one.

**Comment by behdad** [comment]
The same thing happened at OTVar1.8 time: MS insisted that bbox variations must be encoded explicitly and forced it through in HVAR/VVAR tables despite objections; only later it was realized that those cannot be encoded given the existing variation model. It's still hanging over our head. Why is that issue being ignored, even in the face of the more problematic PaintRotate?
